src/python/pants/engine/target.py

```diff
@@ -264,10 +264,74 @@ def __eq__(self, other: Union[Any, AsyncFieldMixin]) -> bool:
         )
 
 
+@union
+@dataclass(frozen=True)
+class FieldDefaultFactoryRequest:
+    """Registers a dynamic default for a Field.
+
+    See `FieldDefaults`.
+    """
+
+    field_type: ClassVar[type[Field]]
+
+
+# TODO: Workaround for https://github.com/python/mypy/issues/5485, because we cannot directly use
+# a Callable.
+class FieldDefaultFactory(Protocol):
+    def __call__(self, field: Field) -> Any:
+        pass
+
+
+@dataclass(frozen=True)
+class FieldDefaultFactoryResult:
+    """A wrapper for a function which computes the default value of a Field."""
+
+    default_factory: FieldDefaultFactory
+
+
+@dataclass(frozen=True)
+class FieldDefaults:
```

Comment:
What about calling it `DynamicFieldDefaults`?

Reply:
I think that all imperative code is implicitly "dynamic"... I wanted to mention dynamic in the docstring to differentiate from static `Field` defaults, but we wouldn't want to call those "static_default", so not using it in the name here seems ok.